build: refactor microbenchmark weekly

herkolategan opened the pull request:
Previously, `roachprod`, `roachprod-microbench` and the microbenchmark weekly
binaries were built using the `dev` tool on CI. It has been updated to rather
use `run_bazel`, which conforms more closely to how other CI jobs build
components.

Additionally, binaries are now identified by its git revision and package
selection. These unique attributes allow us to cache the builds on GCS. The job
will first check if an available build already exists in the specified bucket.
Using GCS to cache the builds also allow us to distribute the binaries from
there instead of copying it through roachprod, when running the microbenchmarks.

Before this change a base set of binaries had to be built beforehand, but the
script now has the ability to build any number of revisions required for the
comparison. This effectively allows anyone to compare any two revisions for
inspection, without having to build a set of binaries by hand.

Release note: None
Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/123882)
<!-- Reviewable:end -->

herkolategan committed:
roachprod-microbench: add compress command

Currently, CI unfortunately lacks more optimised versions of compression tools,
like multithreaded gzip. Due to the size of the microbenchmark binaries and the
level of repetition in those binaries, we require a fast and effective tool to
compress these on the fly. This makes copying and distributing the binaries to a
roachprod cluster more efficient.

This change adds a compress command to allow the CI jobs to utilise `pgzip`
without having it installed on the system.

Release note: None
Epic: None

herkolategan committed:
roachprod-microbench: add stage command

Previously, copying and extracting the binaries were done as part of the run
command. To make CI more flexible and allow the usage of binaries cached on GCE,
a new stage command has been added. The command can stage binaries from both a
local or cloud bucket source. It will extract the binaries, create the necessary
directors and clean up afterward.

Release note: None
Epic: None

herkolategan committed:
roachprod-microbench: managed instance group support

Previously, spot instance support was added by discarding workers where
transient failures have occurred. A new flag `spot` has been added to allow
disabling this behaviour for managed instance groups. In the event of a
preemption on managed instance groups we want to continue retrying, at least for
some period, for the case where the VM is recovered.

Release note: None
Epic: None

herkolategan committed:
roachprod-microbench: decouple remote binaries

Previously, the run command could optionally copy and extract binaries. This
change removes that functionality. It also updates the binaries flags to specify
the location of the remote working directories rather than the local archives.

This allows us to fully decouple the run command from the cluster
initialisation. The CI job can now stage binaries as required, and the run
command will accept the location of these binaries to run the microbenchmarks
from.

Release note: None
Epic: None

herkolategan committed:
build: refactor microbenchmark weekly

Previously, `roachprod`, `roachprod-microbench` and the microbenchmark weekly
binaries were built using the `dev` tool on CI. It has been updated to rather
use `run_bazel`, which conforms more closely to how other CI jobs build
components.

Additionally, binaries are now identified by its git revision and package
selection. These unique attributes allow us to cache the builds on GCS. The job
will first check if an available build already exists in the specified bucket.
Using GCS to cache the builds also allow us to distribute the binaries from
there instead of copying it through roachprod, when running the microbenchmarks.

Before this change a base set of binaries had to be built beforehand, but the
script now has the ability to build any number of revisions required for the
comparison. This effectively allows anyone to compare any two revisions for
inspection, without having to build a set of binaries by hand.

The script has also been updated to use managed instance groups on GCP, so that
it can recover lost spot instances.

Release note: None
Epic: None

renatolabs commented on build/teamcity/cockroach/nightlies/microbenchmark_build_support.sh:
```diff
@@ -0,0 +1,76 @@
+# Check if a given GCS path exists
+function check_gcs_path_exists() {
+  local path=$1
+  if gsutil ls "$path" > /dev/null 2>&1; then
+    return 0
+  else
+    return 1
+  fi
```
bash trivia: this can be rewritten as:

```bash
gsutil ls "$path" &>/dev/null
return
```

herkolategan replied:
Nice, I'll shorten it.

renatolabs commented on build/teamcity/cockroach/nightlies/microbenchmark_build_support.sh:
```diff
@@ -0,0 +1,76 @@
+}
+
+# Build and copy binaries, for the given SHA, to GCS bucket
+function build_and_copy_binaries() {
```
Maybe `upload` is a slightly better name than `copy`?

herkolategan replied:
Good point, I'll rename.

renatolabs reviewed (approved):
This is a great improvement!

My questions are mostly for my own understanding or minor suggestions. LGTM.

renatolabs commented on pkg/cmd/roachprod-microbench/main.go:
```diff
@@ -191,6 +192,19 @@ func makeExportCommand() *cobra.Command {
 
+func makeCompressCommand() *cobra.Command {
+	runCmdFunc := func(cmd *cobra.Command, commandLine []string) error {
+		return compress(cmd.OutOrStdout(), cmd.InOrStdin())
```
Was there a reason this was forced to be stdin and stdout? The automation script needs to redirect stdout to a file. If anyone attempts to use this command locally without knowing about this, it will print random bytes to their terminal, which is never pleasant. Maybe be worth it to accept input and output files.

herkolategan replied:
The main reason for `stdin` is I want to have this pipe directly from `tar` to avoid an interim large file. But as for `stdout` could likely be set to point to a file, but I'd like to keep it simple, since this will most likely ever be used by CI.

herkolategan commented on pkg/cmd/roachprod-microbench/main.go:
```diff
@@ -122,7 +123,28 @@ func makeRunCommand() *cobra.Command {
+
+func makeStageCommand() *cobra.Command {
+	runCmdFunc := func(cmd *cobra.Command, commandLine []string) error {
+		args, _ := splitArgsAtDash(cmd, commandLine)
```
True, can be removed. Done.

herkolategan commented on pkg/cmd/roachprod-microbench/main.go:
```diff
@@ -122,7 +123,28 @@ func makeRunCommand() *cobra.Command {
+		Short: "Stage a given test binaries archive on a roachprod cluster.",
+		Long: `Copy and extract a portable test binaries archive to all nodes on the specified cluster and destination.
 
+The destination should be a directory where the binaries archive will be extracted.
```
True, I'll update the comment.

herkolategan committed:
roachprod-microbench: remove superfluous splitArgsAtDash

Only the `run` command takes args after the dash as extra test args, so we can
remove this split for the other commands.

Epic: None
Release note: None

herkolategan commented on pkg/cmd/roachprod-microbench/stage.go:
```diff
@@ -0,0 +1,77 @@
+	if err != nil {
+		return err
+	}
+	if _, err = roachprod.Sync(l, vm.ListOptions{}); err != nil {
```
Nope, since `roachprod` is generally called before this we don't really need a sync. I'll update.

herkolategan commented on pkg/cmd/roachprod-microbench/stage.go:
```diff
@@ -0,0 +1,77 @@
+	}
+
+	// Remove the remote archive now that it has been extracted.
+	if err = roachprodRun(cluster, l, []string{"rm", "-rf", archiveRemotePath}); err != nil {
```
Yes, good idea. I'll update.

herkolategan commented on pkg/cmd/roachprod-microbench/stage.go:
```diff
@@ -0,0 +1,77 @@
+
+// stage copies the specified archive to the remote machine and extracts it to
+// the specified directory (creating it if it does not exist).
+func stage(cluster, archivePath, remoteDest string) error {
```
Good point, I'll add some context.

herkolategan commented on pkg/cmd/roachprod-microbench/executor.go:
```diff
@@ -103,7 +108,20 @@ func newExecutor(config executorConfig) (*executor, error) {
+	runOptions := install.DefaultRunOptions().WithRetryDisabled()
+	if !config.spot {
+		// If we are not running on spot instances we'll want to retry operations,
+		// for transient errors, in case of a preemption. But since it's possible a
```
I agree, originally I called this flag 'mig', but it also didn't feel right. Maybe something that indicates it will use a long retry mechanism, or that it runs on recoverable VMs.

herkolategan commented:
TFTR!

bors r=renatolabs
